include/ginkgo/core/solver/direct.hpp

```diff
@@ -0,0 +1,143 @@
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_SOLVER_DIRECT_HPP_
+#define GKO_PUBLIC_CORE_SOLVER_DIRECT_HPP_
+
+
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/factorization/factorization.hpp>
+#include <ginkgo/core/solver/lower_trs.hpp>
+#include <ginkgo/core/solver/solver_base.hpp>
+#include <ginkgo/core/solver/upper_trs.hpp>
+
+
+namespace gko {
+namespace experimental {
+namespace solver {
+
+
+template <typename ValueType, typename IndexType>
+class Direct : public EnableLinOp<Direct<ValueType, IndexType>>,
+               public gko::solver::EnableSolverBase<
+                   Direct<ValueType, IndexType>,
+                   factorization::Factorization<ValueType, IndexType>>,
+               public Transposable {
+    friend class EnablePolymorphicObject<Direct, LinOp>;
+
+public:
+    using value_type = ValueType;
+    using index_type = IndexType;
+    using factorization_type =
+        factorization::Factorization<value_type, index_type>;
+    using transposed_type = Direct;
+
+    std::unique_ptr<LinOp> transpose() const override;
+
+    std::unique_ptr<LinOp> conj_transpose() const override;
+
+    GKO_CREATE_FACTORY_PARAMETERS(parameters, Factory)
+    {
+        std::shared_ptr<const LinOpFactory> factorization;
+
+        parameters_type& with_factorization(
+            std::shared_ptr<const LinOpFactory> factory)
+        {
+            this->factorization = std::move(factory);
+            return *this;
+        }
+    };
+    GKO_ENABLE_LIN_OP_FACTORY(Direct, parameters, Factory);
+    GKO_ENABLE_BUILD_METHOD(Factory);
+
+    /** Creates a copy of the solver. */
+    Direct(const Direct&);
+
+    /** Moves from the given solver, leaving it empty. */
+    Direct(Direct&&);
+
+    Direct& operator=(const Direct&);
+
+    Direct& operator=(Direct&&);
+
+protected:
+    Direct(std::shared_ptr<const Executor> exec);
+
+    Direct(const Factory* factory, std::shared_ptr<const LinOp> system_matrix);
+
+    void apply_impl(const LinOp* b, LinOp* x) const override;
+
+    void apply_impl(const LinOp* alpha, const LinOp* b, const LinOp* beta,
+                    LinOp* x) const override;
+
+private:
+    using lower_type = gko::solver::LowerTrs<value_type, index_type>;
+    using upper_type = gko::solver::UpperTrs<value_type, index_type>;
+
+    std::unique_ptr<lower_type> lower_solver_;
+    std::unique_ptr<upper_type> upper_solver_;
```

Comment:
Just from the language side, this assumes that the direct solver uses some form of LU/Cholesky decomposition. Technically, a QR decomposition would also fit as a direct solver. But the naming of the two factors is ill-suited in that case. Naming-wise, I think using `left_solver_` and `right_solver_` might be better. I know that we are not considering that case, so this is just a remark that should at some point be noted somewhere.

Reply:
Great point! Do you think Factorization should also be able to store things like QR factorizations/SVD decompositions or similar? I think that may overload them too much. Are sparse QR solvers of practical importance?

Reply:
I think Factorization would get too overloaded, atm I feel like it's right at the border. I don't know if sparse QR is relevant for us, perhaps @vasilisge0 knows more about that.